extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/tuple/ArrayOfDoublesSketchMergeAggregator.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.query.aggregation.datasketches.tuple;
+
+import com.yahoo.sketches.tuple.ArrayOfDoublesSetOperationBuilder;
+import com.yahoo.sketches.tuple.ArrayOfDoublesSketch;
+import com.yahoo.sketches.tuple.ArrayOfDoublesUnion;
+import io.druid.query.aggregation.Aggregator;
+import io.druid.segment.BaseObjectColumnValueSelector;
+
+public class ArrayOfDoublesSketchMergeAggregator implements Aggregator
+{
+
+  private final BaseObjectColumnValueSelector<ArrayOfDoublesSketch> selector;
+  private ArrayOfDoublesUnion union;
+
+  public ArrayOfDoublesSketchMergeAggregator(final BaseObjectColumnValueSelector<ArrayOfDoublesSketch> selector,
+      final int nominalEntries, final int numberOfValues)
+  {
+    this.selector = selector;
+    union = new ArrayOfDoublesSetOperationBuilder().setNominalEntries(nominalEntries).setNumberOfValues(numberOfValues)
+        .buildUnion();
+  }
+
+  @Override
+  public synchronized void aggregate()
+  {
+    final ArrayOfDoublesSketch update = selector.getObject();
+    if (update == null) {
+      return;
+    }
+    union.update(update);
+  }
+
+  @Override
+  public synchronized Object get()
```

Comment:
See #3956

Reply:
So it seems that we do need to synchronize both aggregate() and get().
For some reason we didn't do that in buffer aggregators. Perhaps, we should as well.

Reply:
Yes, although just using `synchronized` could kill parallelism, because `BufferAggregator` objects are flyweight. I suggest to use `Striped.readWriteLock(64)` and get locks at index that is computed as
```java
int index = (Objects.identityHashCode(buffer) + 31 * offset) & 63;
ReadWriteLock l = striped.get(index);
```

Could you add that?

Reply:
However no, this is all wrong, sorry for confusion.

`synchronized` on `get()` is pointless, because it's not where the returned object is actually used. Something like
```java
synchronized T get() {
  return object.clone();
}
```
would work, but it's bad for performance. What should really be done is some method like `with(Consumer<T>)` added to `Aggregator` interface, but it means overhaul of the whole API and processing engine, definitely not the subject for this PR.

And since `synchronized` on `get()` is pointless, `synchronized` on `aggregate()` is pointless too, because there are no concurrent updates from multiple writers, there is only read (get) / write (aggregate) concurrency.

So as of this PR, I suggest to remove all `synchronized`. This issue should be tackled separately.

Reply:
In this aggregator we either return sketch.compact() or union.getResult(), and both produce another sketch object. So synchronization should help.

Reply:
Please add an elaborate Javadoc comment to each method with `synchronized` modifier in this PR. In those Javadoc comments, please add a link to either this discussion or `https://github.com/druid-io/druid/pull/3956` and explain, how exactly `synchronized` helps to avoid concurrency problems in each particular case (as in here - say about sketch.compact() or union.getResult(), etc.)

Reply:
Added docs and simple synchronization for the buffer aggregators for now. I need to have a closer look at the fine-grained locks. We can lock per buffer position, but we don't know the number of positions upfront, do we?

Reply:
Added striped locks